MSC4161: Crypto terminology for non-technical users

andybalaam opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/andybalaam/crypto-terminology/proposals/4161-crypto-terminology.md)

Conflict of Interest declaration: I am employed by Element. This MSC was written as part of my work on the Element Cryptography team.

----

SCT Stuff

[MSC checklist](https://github.com/matrix-org/matrix-spec-proposals/pull/4161#issuecomment-4114085696)

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/4161#issuecomment-5254578690)

andybalaam committed:
Initial commit for MSC4161: Crypto terminology for non-technical users

andybalaam committed:
First draft

andybalaam committed:
Fix typo

andybalaam committed:
Fix mis-formatted heading

andybalaam committed:
Fix incorrect bolding

andybalaam committed:
Ensure we are consistently saying Alice's identity appears to have changed

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+### Devices
+
+Instances of a client are called 'devices' (not 'sessions'). Aligned with
+MSC4153, we take it as granted that all devices have been cross-signed by the
```
```suggestion
[MSC4153](https://github.com/matrix-org/matrix-spec-proposals/pull/4153), we take it as granted that all devices have been cross-signed by the
```

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+When an unverified person resets their cryptographic identity, we should warn
+the user, so they are aware of the change.
+
+If a verified person's cryptographic identity changes (i.e. they reset their
+recovery key) then this is very important: we verified them because we care
```
As a technical point of detail: it is not necessarily the reset of their recovery key that means their identity changes, but rather their master cross-signing key.

It is technically possible to replace your master cross-signing key without changing your recovery key, and vice-versa. It just happens to be that the most likely reason to rotate your MSK is because you've lost your recovery key.

```suggestion
If a verified person's cryptographic identity changes (i.e. they reset their
master cross-signing key) then this is very important: we verified them because we care
```

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+When an unverified person resets their cryptographic identity, we should warn
+the user, so they are aware of the change.
+
+If a verified person's cryptographic identity changes (i.e. they reset their
+recovery key) then this is very important: we verified them because we care
+about proof that no-one is listening, and now someone could be. The user can
+choose to **withdraw verification** (i.e. "demote" them from being verified), or
+**re-verify** them. Until they do one or the other, communication with them should
+contain a prominent and serious warning that the user's **verified identity has
+changed**.
```
a lot of confusion about which "user"/"person" we are talking about. Rephrase as "Alice/Bob"?

andybalaam replied:
I made an attempt in de285aa0

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+
+⚠️ Avoid talking about "mismatch" or "verification mismatch" which is very
+jargony - it is the identity which is mismatched, not the verification process.
+Just say "Bob's verified identity has changed".
```
```suggestion
Just say "Bob's [verified] identity has changed".
```

may not be verified

clokep commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+Ideally, encryption in Matrix should be entirely invisible to end-users (much as
+WhatsApp or Signal users are not exposed to encryption specifics) - as outlined
+in [MSC4153](https://github.com/matrix-org/matrix-spec-proposals/pull/4153).
+However, consistent terminology is still required, especially during the
+transition period to Invisible Cryptography.
```
MSC4153 is no longer called "Invisible Cryptography" -- is this simply referring to that MSC or is there something else? If something else, this really needs a definition or link.

pmaier1 replied:
[MSC4153](https://github.com/matrix-org/matrix-doc/pull/4153) has been renamed to "Exclude non-cross-signed devices" to make the intention clear. We refer to Invisible Crypto is an umbrella term to summarize quite a couple of things, including this MSC. It also includes other efforts to make crypto more 'invisible' to users like [MSC4048](https://github.com/matrix-org/matrix-spec-proposals/pull/4048), [MSC4147](https://github.com/matrix-org/matrix-spec-proposals/pull/4147) and [MSC4153](https://github.com/matrix-org/matrix-doc/pull/4153).

andybalaam replied:
Updated in a4b95b46

clokep commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+We hope that this MSC will:
+
+* Cause small changes in the spec (as described in the previous paragraph), and
+* Become a document, blessed by the spec core team and available on the spec web
+  site, but probably separate from the spec proper.
```
This is pretty vague -- maybe an appendix would be appropriate?

andybalaam replied:
I like it. Proposed in 0499900a

ara4n commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+changes could be due to attack rather than because they reset their encryption
+(plus "encryption" is jargony).
+
+### Message key
```
We're missing the official terminology for UTDs, and this feels like a great place to specify it (rather than UISIs or UTDs) etc. e.g. "Waiting for message" or "No access to decrypt this message" or whatever.

pmaier1 replied:
Yup, indeed. I'd suggest "Waiting for this message" with a "Learn more" guiding to a FAQ page that explains why users see this and gives some guidance.

andybalaam replied:
First draft added in b919630c

greentore commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,276 @@
+When communicating about cryptography with non-technical users, we propose using
+the following terms and concepts.
+
+### Devices
```
I really don't like calling sessions devices. It may work well for other, mobile first networks, but with matrix it's not all that uncommon to run multiple sessions on the same device.

uhoreg replied:
The problem is that "sessions" is used to refer to other things in the spec.  Also, the Matrix spec already refers to things such as "device IDs" that use the "device" terminology.

MTRNord replied:
How about explicitly naming them "Crypto-Sessions" or something in this direction? I agree with the initial mention about devices. This is already these days due to some UIs a bit confusing.

andybalaam replied:
My personal take is that most people are familiar with "devices", and non-technical users will probably only have one crypto session per device. Further, where people do have multiple sessions on a device I think is it relatively easy to explain that they look like multiple devices to Matrix even though they exist on the same physical device.

In contrast, when I started using Matrix I had no idea what a "session" was, and as I learned more I became further confused because the word is used for several other concepts, including e.g message keys.

So I plan to propose this as-is.

TheArcaneBrony replied:
as much as i hate the "devices" terminology, this is also how it's named over on eg. Discord.

pmaier1 replied:
> as much as i hate the "devices" terminology, this is also how it's named over on eg. Discord.

or Whatsapp or Signal or other platforms. Which is why we suggest it in the first place.

> Has user research shown that devices are preferred over sessions? Does Element have some data on their existing transition from devices to sessions?

Not sure whether we have something dedicated to exactly this comparison. @americanrefugee might know more.

afranke replied:
Gmail uses “session”. I’m sure others do too. I don’t think it’s such a foreign term for laypeople, or that it can’t be understood by them.

![image](https://github.com/user-attachments/assets/436cce4d-dd53-4cec-a6e5-bdf66c9b57d3)

Whatsapp and Signal are more or less tied to phones and don’t expect any other client than “the one app” you have on said phones. In their context, device works. Comparing to them doesn’t work.

afranke replied:
> The problem is that "sessions" is used to refer to other things in the spec. So it would be more confusing to call something else "sessions".

This MSC is about terms to use for non-technical users as far as I can see. I reckon it is not a problem to use a word for something in a given context, and a different word for the same thing in another context. The spec can call these crypto-sessions as was suggested above, or even devices if further disambiguation is deemed necessary. In a technical context, it can be clear that a “device” is not a piece of equipment, whereas it is really not clear for non-technical users.

Without even thinking about multiple distinct clients on the same device, a single client can become a new session if access to the previous session is lost for some reason. And so now you have two “devices” listed even though you used a single computer and a single client.

andybalaam replied:
Element's product people favour "devices", as do Discord, Signal and WhatsApp. We have a counter-example in Google, but I personally find much of Google's UI quite confusing, especially gmail.

As for having two devices listed even though you only used one: the proposed wording seems helpful to me - the mismatch between the wording and reality indicates a problem: they should remove the old device.

Perhaps the product designers from other products could chip in, but at the moment the scale seems to be tipping towards "devices".

richvdh replied:
I think it's worth pointing out that the spec and protocol *already* uses the term "device" for this concept. See in particular:
 * https://spec.matrix.org/v1.13/#devices which discusses the fact that there is not a 1:1 relationship between Matrix devices and physical devices
 * The fact that `/login` returns a `device_id`
 * The fact that [send-to-device messages](https://spec.matrix.org/v1.13/client-server-api/#send-to-device-messaging) are called that.
 
[@andybalaam: we could probably do with an Alternatives section in the MSC, which talks a bit more about why we're recommending "device" rather than "session": that would be a good place to include some of those links].

andybalaam replied:
Added an alternatives section in c13bd45e

andybalaam committed:
Add another example for key storage

andybalaam committed:
Link to MSC4153 where we mention it.

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

andybalaam committed:
Make the wording about resetting keys more precise.

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

andybalaam committed:
Re-word the paragraph about the importance of verified identity changes to use Alice and Bob

andybalaam committed:
Link to the spec equivalent wording for our recovery key

andybalaam committed:
Re-word the paragraph about Invisible Crypto

andybalaam committed:
Propose being an appendix to the spec

andybalaam committed:
Add a section on what to say when a message can't be decrypted

andybalaam committed:
Add a missing 's'

andybalaam committed:
changes->changed

andybalaam committed:
Remove erroneous UTD message

Saiv46 commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+
+### Message key
+
+A **message key** is used to decrypt a message. The metaphor is that messages
+are "locked in a box" by encrypting them, and "unlocked" by decrypting them.
```
Better term could be **encryption key**, as it's more common in news.

dkasak replied:
The problem is there are loads of things in Matrix that could be called an "encryption key". This is specifically a "message encryption key", shortened to "message key".

Saiv46 commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+
+> "Allow key storage"
+
+> "Key storage holds your cryptographic identity on the server along with the
+> keys that allow you to read your message history."
```
> ⚠️ Avoid talking about "cryptographic identity" which is very jargony.

andybalaam replied:
Hmm, yeah. Just "Key storage holds your identity"? Or maybe "Key storage holds your identity information"? @pmaier1 any thoughts?

pmaier1 replied:
> Key storage holds your identity

Yeah, keep it simple, makes sense to me.

dkasak replied:
See https://github.com/matrix-org/matrix-spec-proposals/pull/4161/files#r1778410962 for my concerns about this. "*Key* storage holds your *identity*" sounds extremely confusing to me if I imagine myself as a non-technical newcomer to Matrix.

Saiv46 commented:
This should be given more priority. User-friendliness of clients is a must

turt2live commented on proposals/4161-crypto-terminology.md:
Overall review: The MSC appears to describe features and requirements of Element, and assumes that all clients in the ecosystem have similar or comparable features/needs. The assumptions read fine when applying this MSC to an Element client, but when comparing to the spec there feels like there's a gap in understanding. 

Significant review from other client authors/developers, and their respective product/design teams, I think would be extremely valuable to help ensure this MSC does not solely apply to Element. This kind of review may need to be chased down as it's somewhat uncommon for a product-y MSC to be in the spec process.

turt2live replied:
I don't think this comment is helpful at this stage in the proposal, so resolving. 

We should continue trying to solicit opinions from other developers/teams, but appear to be getting limited responses so far.

turt2live reviewed (changes requested):
This MSC is something I think would be extremely helpful in the spec. It's important to me that users get a relatively consistent experience regardless of the client they choose to use, but still allowing those clients to express their unique features/competitive advantage to users. Clients should also have capacity to participate in different markets outside of chat.

The MSC also appears very nearly ready for FCP, with only a few blocking comments from my side:
* The MSC could do with a bit stronger rationale for *why* this is important. 
* The MSC appears to have a heavy focus on Element clients, which is natural given the authors. 
* To help with the "why" of the MSC, "implementation" in the form of user research would help. Ideally from a broad spectrum of users and clients, demonstrating the terms chosen carry well between different markets, audiences, brands, etc.

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+[MSC4153](https://github.com/matrix-org/matrix-spec-proposals/pull/4153), we take it as granted that all devices have been cross-signed by the
+user who owns them, and we call these **devices**.
+
+Devices which have not been cross-signed by the user are considered an error
+state, primarily to be encountered during the transition to MSC4153 and/or due
+to buggy/incomplete/outdated clients. These devices are referred to as **not
+secure** and presence of them are considered a serious and dangerous error
+condition, similar to an invalid TLS certificate.
+
+> "This device is not secure. Please verify it to continue."
+
+> "Ignoring 5 messages that were sent from a device that is not secure."
+
+> "Confirm it's you" (when asking to verify a device during login)
+
+⚠️ Avoid saying "secure device". All devices are considered secure by default;
+the user doesn't typically need to worry about the fact that insecure devices
+are a thing, given they should only ever occur in error (or transitional)
+scenarios.
+
+⚠️ Avoid saying "trusted device" or "verified device". Devices are not people,
+and it is helpful to use different language for people vs. devices. (However, we
+do use the verb "verify" to describe how to make a device secure. By using the
+same verb, we help users understand the confusing fact that verifying devices
+and verifying people are similar processes, but with different outcomes.)
+
+⚠️ Avoid using "cross-signing", which requires a deeper understanding of
+cryptography to understand.
+
+⚠️ Avoid mentioning "device keys" - a device is just secure or not.
+
+⚠️ Avoid "session" to mean device. Device better describes what most people
+encounter, and is more commonly used in other products.
```
Can we be explicit that "insecure device" (rather than "not secure device") is permitted?

andybalaam replied:
Fixed in 3c7695f2. Originally I avoided "insecure" because I thought it might cause confusion with e.g. a device that was out of date in terms of security fixes, but in retrospect I think that potential confusion exists either way, and avoiding "insecure" made some of the language clunky.

dkasak commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+and is particularly used to describe messages that are stored on the server
+rather than your device(s)
+
+### Key storage
```
I would call this "Secret storage".

The mechanism in question can store arbitrary data, not all of which are keys. The next paragraph already has to perform significant linguistic gymnastics in order to explain that the user's identity is also stored in the *key* storage, even though it is called an *identity*, not a *key*. (The fact that it is a key is an unimportant technical detail from a user perspective and is very jargony.)

On the other hand, explaining that a key is a certain type of secret should be very natural.

pmaier1 replied:
> I would call this "Secret storage".

We discussed this. The challenge we see is that it could be mixed up. It could either mean
1. a storage for secrets
2. a storage that is secret

That's why we discarded it and propose "Key storage". For a user it is irrelevant if something is a secret or a key or a token or a hash or whatnot, technically speaking. From a users point of view it should be a safe place that holds _keys_ to open up stuff on new devices or in case of an emergency.

dkasak replied:
Also, the problem will be further exacerbated if/when new non-key things start getting stored in secret storage.

Xiretza replied:
> We discussed this. The challenge we see is that it could be mixed up. It could either mean
> 
>     1. a storage for secrets
> 
>     2. a storage that is secret

Wouldn't "Secrets storage" fix that problem?

andybalaam replied:
"Secrets storage" is feels unnatural to say in English. The justification for "key storage" is that the most important things (from a non-technical user's point of view) that are stored in this storage are message keys.

The fact that other things also get stored there doesn't need to prevent us naming it after the most important thing. For example, if I have a school locker for P.E. kit that I commonly call a "kit locker" I think it would be perfectly natural for me to say "I keep my Sony Walkman and Rubik's cube in my kit locker"...

Xiretza replied:
> The fact that other things also get stored there doesn't need to prevent us naming it after the most important thing

It doesn't *need* to, but I think it should, considering that the whole point of this exercise is to make the language used less confusing.

richvdh replied:
It seems like we're now proposing "Key storage" to specifically mean key backup, rather than also lumping 4S into it, which seems like it makes this thread outdated?

andybalaam replied:
Yes, that's right. The lumping-together of those 2 ideas didn't work, because whereas key storage can be set up without user interaction, recovery requires the user to acknowledge and write down the recovery key, so the UI must separate them, meaning we get no benefit from lumping them together.

So now:

key storage = storing message keys on the server

recovery = storing "identity" info on the server, where identity means all the private keys: backup key, cross-signing key, dehydrated device key, etc.

richvdh replied:
LGTM. I think we can resolve this thread.

dkasak commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+Instead, we simply say that both cryptographic identity and message keys are
+stored in key storage.
+
+⚠️ Avoid using "key backup" to talk about storing message keys: keeping things on
+the server is not a "backup", but a reliable, cross-device place where this
+information is stored. The word "backup" implies a redundant way to recover lost
+information, but if the user loses their recovery key, this information is lost.
+Clients and servers may wish to offer additional backup services that provide
+true redundancy and disaster recovery, but key storage is not this.
```
I don't entirely agree with this paragraph as it misses some important nuance. If a user loses their recovery key, they will still have a copy of some (potentially all) keys on their signed in clients. It's entirely possible to run a functioning client without key storage.

> a reliable, cross-device place where this information is stored

Moreover, this sounds like a description fitting of the word "backup".

> The word "backup" implies a redundant way to recover lost information, but if the user loses their recovery key, this information is lost.

And the two clauses here don't sound as a contradiction to me. Backups can certainly be gated behind encryption and authentication. This seems to imply that a backup is only truly a backup if no additional information is required to access it.

To be clear, I agree with phasing out the term "backup" for "storage", but IMO the paragraph above does not provide a good argument for this.

andybalaam replied:
Yep, good points. I had another go here: f654c2e6 - let me know what you think.

dkasak replied:
I think this is much better. Thanks.

dkasak commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+
+⚠️ Avoid "4S" or "quad-S" - these are not descriptive terms.
+
+⚠️ Avoid "private key" - this is an implementation detail and a term with
+specific meaning from cryptography.
```
I don't understand this paragraph in the context of its section. Avoid the term "private key" when talking about... what? Key storage? If so, I've never heard someone refer to "key storage" as "private key".

andybalaam replied:
Thanks, yeah not sure it's useful. Removed in 4bcee330

andybalaam committed:
Mention that clients SHOULD follow this guide and document any exceptions

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+  intended audience is different from most of the spec, meaning different words
+  are used from the main spec body.
+
+Clients may, of course, choose to use different language. Some clients may
+deliberately choose to use more technical language, to suit the profiles of
+their users. This document is aimed at clients targeting non-technical users.
```
I wrote some words in 56b5daf1 - let me know what you think.

richvdh replied:
@turt2live are you happy to resolve this thread?

andybalaam committed:
person->user

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+
+### Verified person
+
+When you verify a person they become **verified**. This means that you have
```
Thanks! Fixed in fba4f5ca

andybalaam committed:
Reword sentence on cross-signing

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+> "You verified this person's identity, but it has changed. Please choose to
+> re-verify them or withdraw verification."
+
+⚠️ Avoid using "cross-signing", which requires a deeper understanding of
+cryptography to understand.
```
Fixed in 9246021e

andybalaam committed:
Re-word the "Waiting for this message" sentence

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Add missing word "message"

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Remove unneeded 'private key' warning

andybalaam committed:
Re-word 'key-backup' paragraph to contrast with export.

andybalaam committed:
Make clear that 'insecure device' is permitted

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,308 @@
+
+**Key storage** means keeping cryptographic information on the server. This
+includes the user's cryptographic identity, and/or the message keys needed to
+decrypt messages.
```
I want to split this into 2 separate concepts: I thought we could run storage of message keys together with storage of identity keys etc (recovery) but recently I realised they a really separate, so we probably need something like this:

* **Room key storage** - store room keys on the server, protected by a room key storage key. This room key storage key can be passed around between devices even if you don't have recovery storage.
* **Recovery storage** - store identity keys and the room key storage key on the server, protected by a recovery code. This allows us to recover both our identity and our room key storage key even if we lose all our devices.

richvdh replied:
It's worth noting that the EX designs (https://www.figma.com/design/qTWRfItpO3RdCjnTKPu4mL/Settings?node-id=1653-29941&node-type=frame&t=LeUe8iJVmzG8srM5-0 etc) refer to "key storage" to cover both of these concepts. I think that's ok; we just need to be clear that "key storage" is used for both *room keys* and *recovery*, and that they are set up separately.

richvdh replied:
> Room key storage - store room keys on the server, protected by a room key storage key.

NB they are *message* keys, not room keys.

pmaier1 replied:
IMO it would be great if we could abstract the details to the user. It should just be one feature from a user POV. Either you want the server to store some information for you and get benefits from it or you don't. I don't see a reason why the user should need to understand the exact concepts behind it.

richvdh replied:
Well, it would be quite a change from the way the apps currently work. Currently EX sets up key backup on registration, but doesn't set up recovery until you actively go through the steps (and EW is even more flexible). That's also why the designs above have separate entries for "Allow key storage" and "set up recovery".

Possibly we could change that, and force the user to set up recovery during the registration/login flow; but it would go against the desire to get people using the app quickly. (Also, this MSC is intended to cover matrix clients in general: even if the Element clients force you to set up recovery at the same time as key backup, other clients may not, and having the terminology available to refer to that situation is important.)

richvdh replied:
I don't really think this is there yet, tbh. See other comments about it, but I feel like it's hard to set this out sensibly without making it clear that there are two parts to Key storage: message keys, and recovery. 

Sure the recovery key gives you access to key storage, but only if you have previously set up recovery. We have to be able to talk about "setting up recovery" as a separate step to "enabling key storage".

andybalaam replied:
I did a rewrite in a9383723 . Better?

richvdh replied:
Will close this thread in favour of others.

tulir commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,308 @@
+### Recovery key (and recovery passphrase)
+
+A **recovery key** is a way of regaining access to key storage if the user loses
+all their devices. Using key storage, they can preserve their cryptographic
```
For the record, Beeper settled on "recovery code" after considering other options including "recovery key", but I don't remember the exact reasoning since the switch was made 2 years ago

tulir replied:
Asked for the reasoning internally:

> Code sounded more friendly than key (in an unscientific review).
> e.g. people know what a pin code is. Code is a numerical thing but a 'key' to most people is a physical object

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,308 @@
+proof that you have a secure communication channel with them.
+
+> "Warning: Alice's identity appears to have changed" (when a non-verified
+> user resets their recovery key)
```
We have been trialing the "appears to have changed" wording in Element clients, and it is not popular at all, because it suggest we are unsure what has happened.

The intention of the wording was to make clear that although we are sure that the _cryptographic_ identity has changed, we can't be sure whether this is because the real person did a reset, or some malicious actor is impersonating them.

But this is clearly not working, and we need to work on different wording.

andybalaam replied:
I wonder whether "Alice's identity has changed: there is a chance someone took over their account - to be sure, you should verify their identity in their profile" or something else action-orientated would be helpful.

BillCarsonFr replied:
I think the chances that someone took over the account are in practice very low, and it is alarming to say it up-front. I'd go with `Alice probably re-installed the app and lost her recovery key, you can always verifiy to ensure the identity is legitimate`

richvdh replied:
+1 to "Alice's identity has changed". I think most people will get that. Especially if we can couple it with "there is a chance someone took over their account".

dkasak replied:
Why not copy what other messengers do?

> Alice's identity has changed. [This can happen if the user lost all their devices and the recovery key, but it can also be a sign of someone taking over the account].

Clients are suggested to always display something to the effect of the first sentence. The rest (what's in the brackets) should be available but can be hidden behind a "Tap for more info." interaction or similar.

ara4n replied:
+1 to "Alice's identity has changed." I do think we should be action-focused too, though, so: "Please verify them".

dkasak replied:
Note that this thread is specifically about the non-verified case. It makes sense to mention verification but probably in a non-binding way. I still think there is room for a short single sentence explanation of why this happened, especially if hidden behind an interaction.

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,308 @@
+tooltip or help link): e.g. "Your recovery key may also have been referred to as
+a security key in the past"
+
+⚠️ Be aware that at time of writing the spec uses
+["recovery key"](https://spec.matrix.org/v1.8/client-server-api/#recovery-key)
+to refer to the private half of the backup encryption key, which is different
+from the usage here. The recovery key described in this section is referred to
```
This got fixed in https://github.com/matrix-org/matrix-spec/pull/1819

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,308 @@
+state, primarily to be encountered during the transition to MSC4153 and/or due
+to buggy/incomplete/outdated clients. These devices are referred to as **not
+secure** or **insecure** and their existence is considered a serious and dangerous error
+condition, similar to an invalid TLS certificate.
```
Proposed new wording: "Encryption login" - your device is insecure because you have not completed the encryption login.

andybalaam replied:
![encryption-login](https://github.com/user-attachments/assets/dda81d99-636f-4a85-9e1a-19dc4f928894)

Mock-up of how this might be communicated to user.

andybalaam replied:
Hmm, lots of push-back on this wording in our team, but we like the diagram.

We are playing around with something like "verify it's you".

andybalaam committed:
Remove 'cryptographic' before 'identity', and also tone down the warnings against this

andybalaam committed:
Link to spec sections on QR and emoji, and transitive trust MSC

andybalaam committed:
Add a section about why this is important

andybalaam committed:
Add a paragraph about losing the recovery key

andybalaam committed:
Remove the 'appears to have changed' wording and add explanation

andybalaam committed:
Note that older spec versions say 'recovery key'

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,344 @@
+user who owns them, and we call these **devices**.
+
+Devices which have not been cross-signed by the user are considered an error
+state, primarily to be encountered during the transition to MSC4153 and/or due
```
See the comments in [WorkAdventure's blog post](https://workadventu.re/article-en/managing-e2e-encryption-with-matrix-in-a-simple-way/) under "Compatibility with Element X (and the invisible crypto plan)".

We should make it clear that only unsigned clients _that are doing encryption_ are a problem. Clients that simply don't use any encrypted rooms are fine without being signed.

dkasak replied:
See also https://github.com/matrix-org/matrix-spec-proposals/pull/4153#discussion_r1928557275

richvdh replied:
actually we need to tweak the previous paragraph too.

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+state, primarily to be encountered during the transition to MSC4153 and/or due
+to buggy/incomplete/outdated clients. These devices are referred to as **not
+secure** and presence of them are considered a serious and dangerous error
+condition, similar to an invalid TLS certificate.
```
> Maybe we could include it, but note that it is describing a post-[MSC4153](https://github.com/matrix-org/matrix-doc/pull/4153) world?

It would be much easier if we could just land MSC4153, but in the meantime, yes, this sounds like the best way.

andybalaam replied:
Added a note in d0878852

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,344 @@
+
+### Key storage
+
+**Key storage** means keeping cryptographic information on the server. This
+includes the user's identity, and/or the message keys needed to
+decrypt messages.
```
I think it would be helpful to root this in terms of the underlying protocol:

```suggestion
**Key storage** means keeping cryptographic information on the server. In
terms of the underlying protocol, this consists of two parts: [server-side
key backups](https://spec.matrix.org/v1.13/client-server-api/#server-side-key-backups), and
[server-side key storage](https://spec.matrix.org/v1.13/client-server-api/#key-storage).
```

andybalaam replied:
This is not directly relevant after my rewrite. Did you still want to relate Key Storage with "server-side key backup" and Recovery with "server-side secret storage (4S)"?

richvdh replied:
I think it might be helpful, yes; doing so will (a) help readers familiar with the current paradigms; (b) help people correlate these recommendations to the concepts in the spec.

andybalaam replied:
Done in ecfd1ad4

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,344 @@
+
+> "Message history is unavailable because key storage is disabled."
+
+⚠️ Avoid distinguishing between "secret storage" and "key backup" - these are
+both part of key storage.
```
This feels somewhat confusing as written.

Basically, "secret storage" and "recovery" are the same mechanism, and the Element clients absolutely do distinguish between "Key storage" and "recovery". 

I think what we probably want to say here is "avoid the terms 'key backup' and 'secret storage'".

andybalaam replied:
This is not relevant since my re-write.

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,344 @@
+A **recovery passphrase** is an easier-to-remember way of accessing the recovery
+key and has the same purpose as the recovery key.
+
+**Losing the recovery key**: if the user loses their recovery key, they can
+"reset" it, which means re-storing the identity information in the server,
+encrypted with a new recovery key. If the user has a verified client, then that
+is holding the identity information locally, so they can reset their recovery
+key without losing access to key storage. If they don't have a verified client
+and they lose their recovery key, then they need to reset key storage as well as
+recovery key (since the identity information is needed to read from key
+storage), meaning they lose access to old messages.
```
We now have text in both the "Key storage" and the "Recovery key" sections which talk about what happens if the user loses the recovery key, and they say slightly different things. I think we need to rationalise the two.

andybalaam replied:
I think I addressed this in the rewrite a93837230

andybalaam committed:
Add an Alternatives section about device and session

andybalaam committed:
Add a note about the Devices section depending on MSC4153

andybalaam committed:
Re-write key storage and recovery

andybalaam committed:
Rename 'reset recovery' to 'change recovery key'

andybalaam committed:
Remove recommendation against recovery code

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -83,6 +83,13 @@ clients SHOULD document how they deviate from this document, and why.
 
 ### Devices
 
+**Note: this section depends on [MSC4153](https://github.com/matrix-org/matrix-spec-proposals/pull/4153)
+- Exclude non-cross-signed devices, which specifies clients should avoid sending
+and receiving encryption info with devices that are not cross-signed by their
+owner ("insecure" devices in our terminology). While MSC4153 remains unmerged,
+the parts of this section relating to insecure devices should be considered
+non-normative.**
```
Are the `**`s supposed to make this bold? They aren't working.

richvdh replied:
oh I see, that "-" is supposed to be a dash, not the start of a list.

```suggestion
**Note: this section depends on [MSC4153 ("Exclude non-cross-signed devices")](https://github.com/matrix-org/matrix-spec-proposals/pull/4153), which specifies clients should avoid sending
and receiving encryption info with devices that are not cross-signed by their
owner ("insecure" devices in our terminology).** While MSC4153 remains unmerged,
the parts of this section relating to insecure devices should be considered
non-normative.
```

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,310 @@
+in the spec as the
+[secret storage key](https://spec.matrix.org/v1.8/client-server-api/#secret-storage).
+
+## Potential issues
```
@turt2live are you happy to resolve this thread?

andybalaam committed:
Use a comma instead of a dash to avoid looking like a bullet point.

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

andybalaam committed:
Only unverified devices that have published keys are an error

andybalaam committed:
Link to the equivalent spec names for key storage and recovery

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+# MSC4161: Crypto terminology for non-technical users
```
Lots to think about here but I wanted to say straight away: thank you - you raise lots of very important points and I think I agree with almost everything you've said.

ajbura replied:
>  I'm quite surprised by the session -> device one

It actually make more sense as "session" as terminology should not be exposed to general user. Session as UI term is more driven from developers use of "session tokens".

Generally explaining UI/UX reasoning is hard but let me break it down with screenshot.
<img width="589" alt="Screenshot 2025-02-12 at 9 25 18 AM" src="https://github.com/user-attachments/assets/e4afc0f4-99d9-4cf1-8b4a-52d9fd4bcb4f" />
whenever user login new client:
-  it should be considered as a login *entry* (not session) (marked in blue) in UI.
- At this point we can't say the login entry is done by *you*. 
- To verify it we should perform some check that the device that is used, is owned by *you*?
- Prompt the user to verify the device identity.

Overall login entries (session) are subset of devices, but verification involves proving ownership of devices to grant secure access.

ajbura replied:
No, you are not wrong. Out of overall cohesive UI, individual login entry should not need to be called as *device*(or anything) as it has data associated with it, which makes itself explanatory.
But when do we call them as device? obviously when we are referencing individual entry in ui. like:
- "A new unverified *device* (login) appears: *xyz (macbook)*" here we should purposely call it a device because of security measures which involves *Device verification* (not session verification). Until we verify, it *is* a new device login in overall secure system.
- "Changing *device* display name" It is better to call it device as the display name contains info that help us associate login with specific device.
- "Deleting choosen *devices*" better to call device as login entries are selected by reading device display names.

andybalaam replied:
Hopefully 00993d93 gives the necessary flexibility to use either word as needed.

andybalaam replied:
> Speaking of this MSC: I would appreciate having more external references
> ...
> Or consider "identity" - I don't know if it's good or bad because I don't see the research behind the choice...

This is a very good point. I have attempted to provide some external references in 38e2f257 - if you like the general style of this I can try to do it for some other words.

In some cases I think what we are doing in the MSC is just standardising words that are already in use, so maybe the references should rather be to existing Matrix clients - what do you think?

andybalaam replied:
> I would suggest that we actually spec the glossary, leaving specific phrases to a lighter "implementation guide" kind of a document.

We could definitely do a split like this if there is support for it. So far I have thought of what we are doing as a glossary, with the words in bold as normative, and the additional words used for explanation of the glossary, not necessarily an implementation guide. So I am not yet personally convinced that we need a split, but maybe we could make it clearer that the examples are for explanation purposes?

andybalaam replied:
> I'm not sure we at SCT have experience and expertise on the subject. Being one of those covering the clients domain area, I may have to go for some research and self-education before I can responsibly make decisions on this. And/or we might have to call someone in who has appropriate background. Fortunately, Matrix is used by some UI-conscious communities (GNOME, in particular) so we might not need to search too far away.

I'd certainly welcome input from people with other expertise - this is so far mainly constructed from observing words already in use in multiple Matrix clients, and with expert input from product management and design professionals within Element. We have made adjustments based on input from other clients and community members too, but have not managed to get a huge amount of engagement.

However, I'd be keen not to add an impossible barrier to this MSC. Even if it needs improvements later (which I agree could be painful) I think it can improve the lives of users soon after it is merged, and if it is never merged we will never see those improvements.

andybalaam replied:
>     1. Mainly to SCT: can we make sure this MSC does not end up a single shot and establish a way to further contribute in the same direction?

I believe https://github.com/matrix-org/matrix-spec-proposals/pull/4270 is an attempt to address this comment by establishing a context within which this MSC can be applied.

andybalaam replied:
> For example: next to others, I'm not a fan of "devices", I really believe it's a misnomer we borrowed from centralised mobile-first platforms. I can still grudgingly accept it if we couldn't find a better alternative. But the only discussed alternative was "session", which I agree is (even) weaker. I was surprised that "clients" (taken from e-mail), or "applications/apps" (desktop and mobile environments) are not even namechecked.

Thanks, I added more alternatives in fa893e15 . I personally don't think there are better words than "device", but we can try! btw I'm not sure how "devices" is related to being centralised? (I do see how it's from mobile-first platforms, and I think we should embrace that a little since it is how so many people engage with messaging.)

richvdh commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+the parts of this section relating to insecure devices should be considered
+non-normative.
+
+Instances of a client are called 'devices' (not 'sessions'). Aligned with
```
A question that arose in discussion with the SCT: do we actually *need* to specify "sessions" vs "devices"? Maybe this is something that we can leave up to individual clients, and users will figure it out?

andybalaam replied:
I updated this MSC here 00993d93 with similar language to the above, with a slightly more even-handed wording. Given the strength of feeling on this issue I don't want to describe 'sessions' as a second choice, but rather an alternative.

dkasak commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+the parts of this section relating to insecure devices should be considered
+non-normative.
+
+Instances of a client are called 'devices' (not 'sessions'). Aligned with
```
There's a thing that's been bugging me about this. Are "devices" and "sessions" even the same thing? MAS is primarily talking about server-side sessions which correspond to access tokens. But when we are talking about cryptography, we don't mean that but something which is exclusively client-side and corresponds to cryptographic key material.

Sure, there is often a correspondence between the two, but they are not the same thing. They are not even necessarily 1-to-1 given there are client instances which have no cryptography enabled.

So is the terminology problem partly caused by our stumbling over this difference without realising it?

andybalaam replied:
Updated in 00993d93 to allow either device or session. It's clear that there is no consensus in the community about which word to use, so we take the slight hit of having 2 words, but we avoid clients having to diverge from this proposal.

As for @dkasak 's comments: I think there might be something very helpful there, but I think it's too low-level for this proposal.

dkasak replied:
@andybalaam Do you have any suggestions where else we can take it? It's a bit of a fear of mine that the wording used in this MSC, should it get accepted, will interfere with the established technical wording in the spec.  Both this MSC and the spec talk about "devices", but they really mean (related but ultimately) different things, don't they?

andybalaam replied:
I actually think devices here are quite close to what the spec means when it says devices. Where do you see a difference?

In general, I'm not sure where we should take your comments. If you want to propose a change to the wording used in the main spec, I suppose a separate MSC would be the best way.

dkasak replied:
I think my complaint is primarily that the existing treatment of devices in the spec is ambiguous and somewhat confused. This MSC purports to improve the terminology but is not actually disentangling this ambiguity but rather doubling down on it.

In particular, we do not properly disentangle between:

1. Physical devices or client installations.
2. Cryptographic devices, i.e. the data structures representing the root of trust and authenticity of a given installation via cryptographic means. This roughly corresponds to the `DeviceKeys` struct in the spec.
3. Server-side or MAS-side sessions.

There are some loose business logic relations between these concepts but they are *distinct* concepts. The current spec treats them somewhat interchangeably, ambiguously referring to all of them as "devices". My argument is that each of these concepts should get its own term and that the relationships between them ought to be unambiguously described using those terms.

turt2live commented on proposals/4161-crypto-terminology.md:
Having discussed this briefly with the SCT, the actual implementation requirements are more along the lines of requiring Element to implement this MSC. This is because of the client's large user base, allowing us to determine if the terms appeal to users in that kind of environment. 

Other clients can and should implement this MSC still, however. Those implementation would further add data to the terms proposed in this MSC, either favourably or otherwise :)

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+#### Logging out
+
+In contrast to some other services, **logging out** (or **signing out**) of a
+Matrix device is quite a significant operation: it means the encryption data on
+this device is lost, and if you log out of all devices you will need to use your
```
```suggestion
Matrix device is quite a significant operation: it means your encryption data on
the device is lost, and if you log out of all devices you will need to use your
```

nit: change of voice/ownership.

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+In many contexts, most users are **not verified**: verification is a manual
+step ([scanning a QR code](https://spec.matrix.org/v1.12/client-server-api/#qr-codes) or [comparing emojis](https://spec.matrix.org/v1.12/client-server-api/#sas-method-emoji)). (In future, verification will
+probably become more common thanks to [MSC2882 Transitive Trust](https://github.com/matrix-org/matrix-spec-proposals/pull/2882) or something similar).
+When an unverified user resets their identity, we should warn
+the user, so they are aware of the change.
```
it's not immediately clear which user we're warning here, and I don't have good suggestions for how to fix the sentence without adding more terminology. 

Examples of what I'm thinking of:

> When an unverified user/contact resets their identity, the user should be warned that this happened.

> When an unverified user resets their identity, all the users they verified should be warned that this happened.

> When an unverified user resets their identity, the observing user should be warned that this happened.

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+When communicating about cryptography with non-technical users, we propose using
+the following terms and concepts.
+
+When referring to concepts outlined in this document in their user interface,
+clients SHOULD use the language specified, except where their own users are
+known to understand different terms more easily. When making such exceptions,
+clients SHOULD document how they deviate from this document, and why.
```
```suggestion
When referring to concepts outlined in this document in their user interface,
clients SHOULD use the language specified in **bold**, except where their own users are
known to understand different terms more easily. When making such exceptions,
clients SHOULD document how they deviate from this document, and why.
```

editorial thing: somewhere the MSC needs to say that bolded terms are normative, and this seems like a fine place.

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+⚠️ Avoid using "room key". These keys are used to decrypt messages, not rooms.
+
+Note: this clashes with the term "message key" in the double ratchet. Since the
+double ratchet algorithm is for a very different audience, we think that this is
+not a problem.
```
```suggestion
double ratchet algorithm is for a very different audience, this is not considered
a blocking conflict.
```

editorial: assertive/proposal language

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+When we expect the key to arrive, we are **waiting for this message**.
+
+> "Waiting for this message" with a button: "learn more" that explains that the message key for
+this message has not yet been received, but that we expect it to
```
```suggestion
> this message has not yet been received, but that we expect it to
```

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+double ratchet algorithm is for a very different audience, we think that this is
+not a problem.
+
+### Unable to decrypt
```
For the MSC to consider: Can we say "unable to decrypt" if there was a technical error during decryption? ie: the message key was received, but it's not quite in the right format, causing an exception to be raised. Or, the message key was received but the client's IndexedDB exploded.

Suggestion: Clients should distinctly identify technical errors from lack of received message key. For example, by saying "something went wrong with this message" instead of "unable to decrypt". "Decrypt" is considered jargon.

andybalaam replied:
Good point - added in cc83a417

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+In the spec, key storage is referred to as
+[server-side key backup](https://spec.matrix.org/v1.13/client-server-api/#server-side-key-backups).
+
+> "Allow key storage"
+
+> "Key storage holds the keys that allow you to read your message history."
+
+> "Message history is unavailable because key storage is disabled."
```
For the MSC to consider: it may be alarming to users that keys aren't stored by default? Adding the context that they are stored *securely* server-side may help ensure that people make an informed decision to turn this on or off.

andybalaam replied:
I'm not yet sure what you are suggesting. Possibilities:

* Because it may be alarming that keys aren't stored by default: we should rename key storage to "server key storage" or similar to make it easier to understand that keys are already stored locally
* Because we want to encourage users that turning on key storage is safe: we should refer to key storage as "secure"

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+messages, and you will be able to read your message history, even if you logged
+out everywhere or lost your devices.
+
+A **recovery passphrase** is an easier-to-remember way of accessing the recovery
+key and has the same purpose as the recovery key.
```
For the MSC to consider: a third term for the same thing is confusing to me as a reader. It's not clear from the description here what "accessing the recovery key" means. What is the thing I put in my password manager?

andybalaam replied:
Thanks - I attempted to address this here: 30acaacb

turt2live reviewed (approved):
Thanks for your continued work on this MSC! It's shaping up to be extremely well written, and may be possible to include verbatim in the spec itself (when we get there). 

This review is a bit large because it contains primarily editorial considerations, but within there is a suggestion to skip the device vs session debate by allowing both, and preferring device. Ultimately, compared to the other terms in this proposal, I don't feel the difference in terminology would measurably add confusion for users. Provided of course that someone doesn't start calling them "apps" or "connections".

andybalaam committed:
Use assertive language rather than "we propose"

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Allow either 'device' or 'session'

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+
+Clients may, of course, choose to use different language. Some clients may
+deliberately choose to use more technical language, to suit the profiles of
+their users. This document is aimed at clients targeting non-technical users.
```
I agree, but I want to make sure the scope of this document is as clear as possible, to avoid unnecessary debate about issues that are more technical than the issues addressed here.

andybalaam committed:
Be more assertive about updating the spec where relevant.

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Be more vague about where this lands in the spec

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Improve explanation of why to avoid 'device keys'

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Tweak comments about loss of devices

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Clarify who should be warned on identity reset

andybalaam committed:
Make clear that bolded terms are normative.

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Add missing quotation prefix

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Proposal language for double ratchet sentence.

Co-authored-by: Travis Ralston <travisr@matrix.org>

andybalaam committed:
Remove MSC4153 material - it has been proposed to be added in MSC4153

andybalaam committed:
Note that Element is using this as a reference

andybalaam commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,451 @@
+We hope that Matrix client developers will like the terms and wording we
+provide, and adapt their user interfaces and documentation to use them. (If this
+MSC is accepted, Element will use it as a reference for English wording in its
+clients.)
```
Updated in 40c268d3

andybalaam committed:
Call out the situation when UTD is because of an error

andybalaam committed:
Explain a little more what a recovery passphrase actually is

andybalaam committed:
Provide a list of references for using the word identity

andybalaam committed:
Add more alternatives to device

andybalaam committed:
Reword 'identity was changed' to 'identity has been reset'

Based on real-world feedback from Element clients, 'identity was changed', and
especially 'identity appears to have been changed' were very confusing.

Resetting an identity is something that the user can do themselves, so it is
much clearer what we mean. They can ask the other user "did you reset your
identity?" and it is easier for them to answer.

escix commented:
Why not use 'client' for 'devices/sessions'.
In reality, each of the 'device/session' is actually a client that we are trying to link with other 'clients'

andybalaam committed:
Talk about confirmation of devices instead of verification

andybalaam committed:
Talk about linking devices

andybalaam committed:
Talk about digital identity instead of just identity

andybalaam committed:
Promote 'confirmed' to a defined term

turt2live commented on proposals/4161-crypto-terminology.md:
```diff
@@ -0,0 +1,464 @@
+don't have access to this message**.
+
+> "You don't have access to this message" e.g. if it was sent before the user
+> entered the room, or the user does not have key storage set up.
```
presumably this also includes when the sender refuses to provide the key (using a withheld error code)?

turt2live reviewed (changes requested):
This MSC is currently in queue for proposed-FCP, though I feel there's a couple more steps that need to happen before it can confidently enter that state. The MSC checklist I'll add shortly after this review will cover the points in more detail.

The things I feel need addressing before this is ready for proposed-FCP are:

* **Resolution of outstanding comments (*prior to this review*).** These threads are either long, old, or both. They should be incorporated into the proposal text or marked as resolved in GitHub if already addressed. 
  * The exception is (always) the "implementation requirements" thread - that takes place of implementations in the description.
* **Evidence of implementation.** See my comment on the "implementation requirements" thread.

When these two points, and any other unchecked boxes from the checklist or concerns from other SCT members, are addressed, please re-raise this MSC for proposed-FCP in the [SCT Office](https://matrix.to/#/#sct-office). Questions regarding threads are best asked in those threads. Questions for the SCT which are outside of existing threads should be raised in the SCT Office (SCT members might not see individual/team pings).

Overall though, in case it's not totally clear from the small number of comments, this proposal is *extremely close* to proposed-FCP. I'm looking forward to it getting pushed to that state!

CrazyNicc reviewed (commented):
Just a minor suggestion for improving the clarity, and using one less term.
